docs/experiments/generic-pattern-matching.md

```diff
@@ -137,6 +137,61 @@ name = ...
   name = ...
 ```
 
+### How to handle line-based input
+
+Many simple configuration formats are a collection of key/value pairs
+delimited by newlines. For example, we would like to extract the
+`password` value from the following made-up input:
+
+```
+username = bob
+password = p@$$w0rd
+server = example.com
+```
+
+Unfortunately, the following pattern won't work the way we want
+because in generic mode, metavariables will only capture a single word
+(alphanumeric sequence):
+
+```
+password = $PASSWORD
+```
+
+This would match the input file but would assign the value `p`
+to `$PASSWORD` instead of the full value `p@$$w0rd`.
+
+To match an arbitrary sequence of items (words, punctuation, etc.) and
+capture its value, we must use a named ellipsis. Our pattern becomes:
+
+```
+password = $...PASSWORD
+```
+
+Unfortunately, this doesn't work either because it captures too
+much. The value assigned to `$...PASSWORD` is now
+`p@$$w0rd`&lt;newline&gt;`server = example.com`. This is because in
+generic mode, an ellipsis extends until the end of the current block
+or up to 10 lines down, whichever comes first. To prevent this,
+specify the option `generic_ellipsis_max_span: 0` in the Semgrep
+rule. This will force the ellipsis to match within a single line.
+The [resulting
+rule](https://semgrep.dev/playground/s/returntocorp:password-in-config-file)
+is:
+
+```yaml
+id: password-in-config-file
+pattern: |
+  password = $...PASSWORD
+options:
+  # prevent ellipses from matching multiple lines
+  generic_ellipsis_max_span: 0
+message: |
+  password found in config file: $...PASSWORD
+languages:
+  - generic
+severity: WARNING
+```
```

Comment:
Use steps to draw the attention of readers, and guide them by actions through hypothetical or real examples. See https://www.notion.so/r2cdev/r2c-technical-communication-style-guide-4d1104e78927460aa8c7624aedd6561f#ea454bcf7ccf4cf38e12247bdb12251f

```suggestion
To match an arbitrary sequence of items and capture their value in the this example:

1. Use a named ellipsis, by changing the pattern to the following:

    ```yaml
    password = $...PASSWORD
    ```

    This still leads Semgrep to capture too much. The value assigned to `$...PASSWORD` are now `p@$$w0rd` and<br />
    `server = example.com`. In generic mode, an ellipsis extends until the end of the current block or up to 10 lines down, whichever comes first. To prevent this behavior, continue with the next step.

2. In the Semgrep rule, specify the following key:
   
    ```yaml
    generic_ellipsis_max_span: 0
    ```

    This option forces the ellipsis operator to match patterns within a single line.
    Example of the [resulting rule](https://semgrep.dev/playground/s/returntocorp:password-in-config-file):

    ```yaml
    id: password-in-config-file
    pattern: |
    password = $...PASSWORD
    options:
    # prevent ellipses from matching multiple lines
    generic_ellipsis_max_span: 0
    message: |
    password found in config file: $...PASSWORD
    languages:
    - generic
    severity: WARNING
    ```

```

This is a preview of rendered output of the suggestion above:
<img width="986" alt="Screenshot 2022-07-29 at 17 47 35" src="https://user-images.githubusercontent.com/26669307/181796415-ba266b27-306d-4de0-aafd-5123887b382a.png">